pkg/endpoint/local_ip.go

```diff
@@ -53,12 +53,24 @@ func getLocalIPFromRoutes(family k8snet.IPFamily) (string, error) {
 }
 
 func GetLocalIPForDestination(dst string, family k8snet.IPFamily) string {
-	conn, err := Dial("udp"+string(family), "["+dst+"]:53")
+	conn, err := Dial("udp"+string(family), net.JoinHostPort(dst, "53"))
+	var ip net.IP
 	if err == nil {
 		defer conn.Close()
 		localAddr := conn.LocalAddr().(*net.UDPAddr)
+		ip = localAddr.IP
 
-		return localAddr.IP.String()
+		if family == k8snet.IPv4 {
+			return ip.String()
+		}
+	}
+
+	if family == k8snet.IPv6 && !isValidGlobalIPv6(ip) {
```

Comment:
Do we want to run this logic even if Dial("udp"+string(family), net.JoinHostPort(dst, "53")) failed ?

Reply:
It seems this logic should be in the if block above since if `err != nil` then `ip` would be nil and `getValidGlobalIPv6FromSameInterface` would be a no-op anyway. Also it might be useful to log a message if the returned ip isn't valid (I assume a warning if it shouldn't normally occur).

```
if err == nil {
    ...
    if family == k8snet.IPv4 {
        return localAddr.IP.String()
    }

    if !isValidGlobalIPv6(localAddr.IP) {
        logger.Warningf("IPv6 %q returned from Dial isn't usable - trying the same interface". localAddr.IP)
        ...
    }
}

// connection failed try fallback method
localIP, err := getLocalIPFromRoutes(family)
...
```